File: supabase/functions/_shared/validate.ts
// Validation of extracted ID fields against the TCAD property record. Pure functions — unit-tested in Deno
// (process-claim/validate_test.ts) and mirrored in Python (trd/agent/validate.py). Both read tests/fixtures/cases.json and
// the snapshot tests/fixtures/findings_snapshot.json pins identical output (G-9).
// Validators emit CODES + facts; every sentence comes from the generated rule table (findings.ts, ADR 0016).
import { blocking, type Finding, makeFinding, reasonText } from "./findings.ts";

export { blocking, reasonText, type Finding };

export type Extracted = {
  readable: boolean; id_type: string; issuing_state: string;
  first_name: string; middle_name?: string; last_name: string;
  dob: string; expiry: string; dl_number: string;
  address_line1: string; city: string; state: string; zip: string;
  confidence: { name: number; dob: number; address: number; dl_number: number; expiry: number };
  issues: string[];
  source?: "photo" | "typed";
};

export type PropertyRec = {
  prop_id: number; owner_name: string; situs_num?: string; situs_street?: string; situs_unit?: string;
  situs_city?: string; situs_zip?: string; situs_full: string; deed_date?: string;
};

export type Validation = {
  status: "ready_to_submit" | "needs_dl_update" | "needs_review";
  address_match: boolean; name_match: boolean; texas_id: boolean; expired: boolean; age: number | null; over65: boolean;
  findings: Finding[];
};

const SUFFIX: Record<string, string> = {
  STREET: "ST", ST: "ST", AVENUE: "AVE", AV: "AVE", AVE: "AVE", BOULEVARD: "BLVD", BLVD: "BLVD", DRIVE: "DR", DR: "DR",
  ROAD: "RD", RD: "RD", LANE: "LN", LN: "LN", COURT: "CT", CT: "CT", CIRCLE: "CIR", CIR: "CIR", PLACE: "PL", PL: "PL",
  TRAIL: "TRL", TRL: "TRL", PARKWAY: "PKWY", PKWY: "PKWY", WAY: "WAY", TERRACE: "TER", TER: "TER", COVE: "CV", CV: "CV",
  LOOP: "LOOP", PASS: "PASS", PATH: "PATH", RUN: "RUN", HIGHWAY: "HWY", HWY: "HWY", NORTH: "N", N: "N", SOUTH: "S", S: "S",
  EAST: "E", E: "E", WEST: "W", W: "W",
};
const UNIT_WORDS = new Set(["APT", "UNIT", "STE", "SUITE", "#", "BLDG", "LOT"]);

export function normalizeStreet(s: string): { num: string; tokens: string[]; unit: string } {
  const raw = (s ?? "").toUpperCase().replace(/[.,]/g, " ").replace(/#/g, " # ").replace(/\s+/g, " ").trim();
  const parts = raw.split(" ").filter(Boolean);
  let num = "", unit = "";
  const tokens: string[] = [];
  for (let i = 0; i < parts.length; i++) {
    const p = parts[i];
    if (!num && /^\d+[A-Z]?$/.test(p)) { num = p.replace(/[A-Z]$/, ""); continue; }
    if (UNIT_WORDS.has(p)) { unit = (parts[i + 1] ?? "").replace(/^#/, ""); i++; continue; }
    tokens.push(SUFFIX[p] ?? p);
  }
  return { num, tokens, unit };
}

export function addressMatches(idLine1: string, idZip: string, prop: PropertyRec): boolean {
  const a = normalizeStreet(idLine1);
  const situsLine = prop.situs_num ? `${prop.situs_num} ${prop.situs_street ?? ""} ${prop.situs_unit ? "UNIT " + prop.situs_unit : ""}` : prop.situs_full.split(",")[0];
  const b = normalizeStreet(situsLine);
  if (!a.num || a.num !== b.num) return false;
  // first street-name token must match; suffix/direction may be dropped on the ID
  const an = a.tokens.filter((t) => !Object.values(SUFFIX).includes(t) || t.length > 2);
  const bn = b.tokens.filter((t) => !Object.values(SUFFIX).includes(t) || t.length > 2);
  if (!an.length || !bn.length || an[0] !== bn[0]) return false;
  const zipA = (idZip ?? "").slice(0, 5), zipB = (prop.situs_zip ?? "").slice(0, 5) || (prop.situs_full.match(/\b(\d{5})\b/)?.[1] ?? "");
  if (zipA && zipB && zipA !== zipB) return false;
  if (b.unit && a.unit && a.unit !== b.unit) return false;
  return true;
}

function nameTokens(s: string): string[] {
  return (s ?? "").toUpperCase().replace(/[^A-Z &]/g, " ").split(/\s+/).filter((t) => t && t !== "&" && t.length > 1);
}

/** TCAD owner names look like "GARCIA RICHARD L", "SMITH JOHN & JANE", "DOE JANE R & JOHN". The applicant's last and first names must both appear. */
export function nameMatches(first: string, last: string, ownerName: string): boolean {
  const toks = nameTokens(ownerName);
  const L = (last ?? "").toUpperCase().replace(/[^A-Z]/g, "");
  const F = (first ?? "").toUpperCase().replace(/[^A-Z]/g, "");
  if (!L || !F) return false;
  const lastOk = toks.some((t) => t.replace(/[^A-Z]/g, "") === L) || toks.join("").includes(L);
  const firstOk = toks.some((t) => t.replace(/[^A-Z]/g, "") === F || (t.length >= 3 && F.startsWith(t)) || (F.length >= 3 && t.startsWith(F)));
  return lastOk && firstOk;
}

export function ageOn(dob: string, on = new Date()): number | null {
  const m = /^(\d{4})-(\d{2})-(\d{2})$/.exec(dob ?? "");
  if (!m) return null;
  const b = new Date(Date.UTC(+m[1], +m[2] - 1, +m[3]));
  let a = on.getUTCFullYear() - b.getUTCFullYear();
  const before = on.getUTCMonth() < b.getUTCMonth() || (on.getUTCMonth() === b.getUTCMonth() && on.getUTCDate() < b.getUTCDate());
  if (before) a--;
  return a;
}

/** `today` defaults to now; tests and the snapshot pass a fixed date so age/expiry are deterministic. */
export function validate(ex: Extracted, prop: PropertyRec, typedName: string, today = new Date()): Validation {
  const findings: Finding[] = [];
  const add = (code: Finding["code"], detail?: Record<string, unknown>) => findings.push(makeFinding(code, detail));
  const texas = (ex.issuing_state ?? "").toUpperCase() === "TX" && ex.id_type !== "other";
  const addr = addressMatches(ex.address_line1, ex.zip, prop);
  const nm = nameMatches(ex.first_name, ex.last_name, prop.owner_name);            // ID vs. owner of record (the eligibility test)
  const signerOk = nameMatches(ex.first_name, ex.last_name, typedName);            // ID vs. the name typed as signature (identity test)
  const age = ageOn(ex.dob, today);
  const expired = !!ex.expiry && ex.expiry < today.toISOString().slice(0, 10);
  const lowConf = Math.min(ex.confidence?.name ?? 0, ex.confidence?.address ?? 0, ex.confidence?.dob ?? 0) < 0.6;
  const idName = `${ex.first_name ?? ""} ${ex.last_name ?? ""}`;
  const idAddress = `${ex.address_line1 ?? ""}, ${ex.city ?? ""} ${ex.zip ?? ""}`;

  if (!ex.readable) add("not_readable");
  if (!texas) add("not_texas_id", { issuing_state: ex.issuing_state, id_type: ex.id_type });
  if (!addr) add("address_mismatch", { id: idAddress, situs: prop.situs_full });
  else add("address_match");
  if (!nm) add("name_mismatch", { id: idName, owner: prop.owner_name });
  else add("name_match");
  if (!signerOk) add("signer_mismatch", { id: idName, typed: typedName });
  if (expired) add("expired", { expiry: ex.expiry });
  if (lowConf) add("low_confidence", { confidence: ex.confidence });
  if (age != null && age >= 65) add("over_65", { age });
  if (age != null && age < 18) add("under_18", { age });

  let status: Validation["status"] = "ready_to_submit";
  if (!ex.readable || !texas || !nm || !signerOk || lowConf || (age != null && age < 18)) status = "needs_review";
  else if (!addr) status = "needs_dl_update";
  return { status, address_match: addr, name_match: nm, texas_id: texas, expired, age, over65: age != null && age >= 65, findings };
}

File: cleanbill/agent/validate.py
"""Python mirror of supabase/functions/_shared/validate.ts — same rules, same outputs.
Keep the two in sync: both read tests/fixtures/cases.json and tests/fixtures/findings_snapshot.json pins identical rendered
findings (G-9). Validators emit CODES + facts; every sentence comes from the rule table in cleanbill/findings.py (ADR 0016).
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import date

from cleanbill.findings import blocking, make_finding, reason_text  # noqa: F401  (re-exported for callers)

SUFFIX = {
    "STREET": "ST", "ST": "ST", "AVENUE": "AVE", "AV": "AVE", "AVE": "AVE", "BOULEVARD": "BLVD", "BLVD": "BLVD", "DRIVE": "DR", "DR": "DR",
    "ROAD": "RD", "RD": "RD", "LANE": "LN", "LN": "LN", "COURT": "CT", "CT": "CT", "CIRCLE": "CIR", "CIR": "CIR", "PLACE": "PL", "PL": "PL",
    "TRAIL": "TRL", "TRL": "TRL", "PARKWAY": "PKWY", "PKWY": "PKWY", "WAY": "WAY", "TERRACE": "TER", "TER": "TER", "COVE": "CV", "CV": "CV",
    "LOOP": "LOOP", "PASS": "PASS", "PATH": "PATH", "RUN": "RUN", "HIGHWAY": "HWY", "HWY": "HWY", "NORTH": "N", "N": "N", "SOUTH": "S", "S": "S",
    "EAST": "E", "E": "E", "WEST": "W", "W": "W",
}
UNIT_WORDS = {"APT", "UNIT", "STE", "SUITE", "#", "BLDG", "LOT"}
SUFFIX_VALUES = set(SUFFIX.values())


def normalize_street(s: str) -> tuple[str, list[str], str]:
    raw = re.sub(r"\s+", " ", (s or "").upper().replace(".", " ").replace(",", " ").replace("#", " # ")).strip()
    parts = [p for p in raw.split(" ") if p]
    num, unit, tokens = "", "", []
    i = 0
    while i < len(parts):
        p = parts[i]
        if not num and re.fullmatch(r"\d+[A-Z]?", p):
            num = re.sub(r"[A-Z]$", "", p); i += 1; continue
        if p in UNIT_WORDS:
            unit = (parts[i + 1] if i + 1 < len(parts) else "").lstrip("#"); i += 2; continue
        tokens.append(SUFFIX.get(p, p)); i += 1
    return num, tokens, unit


def address_matches(id_line1: str, id_zip: str, prop: dict) -> bool:
    a_num, a_tok, a_unit = normalize_street(id_line1)
    if prop.get("situs_num"):
        situs_line = f"{prop['situs_num']} {prop.get('situs_street') or ''} {('UNIT ' + prop['situs_unit']) if prop.get('situs_unit') else ''}"
    else:
        situs_line = prop["situs_full"].split(",")[0]
    b_num, b_tok, b_unit = normalize_street(situs_line)
    if not a_num or a_num != b_num:
        return False
    an = [t for t in a_tok if t not in SUFFIX_VALUES or len(t) > 2]
    bn = [t for t in b_tok if t not in SUFFIX_VALUES or len(t) > 2]
    if not an or not bn or an[0] != bn[0]:
        return False
    zip_a = (id_zip or "")[:5]
    zip_b = (prop.get("situs_zip") or "")[:5] or (re.search(r"\b(\d{5})\b", prop["situs_full"] or "") or [None, ""])[1]
    if zip_a and zip_b and zip_a != zip_b:
        return False
    if b_unit and a_unit and a_unit != b_unit:
        return False
    return True


def _name_tokens(s: str) -> list[str]:
    return [t for t in re.sub(r"[^A-Z &]", " ", (s or "").upper()).split() if t and t != "&" and len(t) > 1]


def name_matches(first: str, last: str, owner_name: str) -> bool:
    toks = _name_tokens(owner_name)
    L = re.sub(r"[^A-Z]", "", (last or "").upper()); F = re.sub(r"[^A-Z]", "", (first or "").upper())
    if not L or not F:
        return False
    last_ok = any(re.sub(r"[^A-Z]", "", t) == L for t in toks) or L in "".join(toks)
    first_ok = any(re.sub(r"[^A-Z]", "", t) == F or (len(t) >= 3 and F.startswith(t)) or (len(F) >= 3 and t.startswith(F)) for t in toks)
    return last_ok and first_ok


def age_on(dob: str, on: date | None = None) -> int | None:
    m = re.fullmatch(r"(\d{4})-(\d{2})-(\d{2})", dob or "")
    if not m:
        return None
    on = on or date.today()
    b = date(int(m[1]), int(m[2]), int(m[3]))
    a = on.year - b.year - ((on.month, on.day) < (b.month, b.day))
    return a


@dataclass
class Validation:
    status: str
    address_match: bool
    name_match: bool
    texas_id: bool
    expired: bool
    age: int | None
    over65: bool
    findings: list[dict] = field(default_factory=list)

    def as_dict(self) -> dict:
        return self.__dict__.copy()


def validate(ex: dict, prop: dict, typed_name: str, today: date | None = None) -> Validation:
    today = today or date.today()
    texas = (ex.get("issuing_state") or "").upper() == "TX" and ex.get("id_type") != "other"
    addr = address_matches(ex.get("address_line1", ""), ex.get("zip", ""), prop)
    nm = name_matches(ex.get("first_name", ""), ex.get("last_name", ""), prop["owner_name"])
    signer_ok = name_matches(ex.get("first_name", ""), ex.get("last_name", ""), typed_name)
    age = age_on(ex.get("dob", ""), today)
    expired = bool(ex.get("expiry")) and ex["expiry"] < today.isoformat()
    conf = ex.get("confidence") or {}
    low_conf = min(conf.get("name", 0), conf.get("address", 0), conf.get("dob", 0)) < 0.6

    id_name = f"{ex.get('first_name') or ''} {ex.get('last_name') or ''}"
    id_address = f"{ex.get('address_line1') or ''}, {ex.get('city') or ''} {ex.get('zip') or ''}"

    findings: list[dict] = []
    add = lambda code, detail=None: findings.append(make_finding(code, detail))  # noqa: E731
    if not ex.get("readable", True): add("not_readable")
    if not texas: add("not_texas_id", {"issuing_state": ex.get("issuing_state"), "id_type": ex.get("id_type")})
    if not addr: add("address_mismatch", {"id": id_address, "situs": prop["situs_full"]})
    else: add("address_match")
    if not nm: add("name_mismatch", {"id": id_name, "owner": prop["owner_name"]})
    else: add("name_match")
    if not signer_ok: add("signer_mismatch", {"id": id_name, "typed": typed_name})
    if expired: add("expired", {"expiry": ex.get("expiry")})
    if low_conf: add("low_confidence", {"confidence": conf})
    if age is not None and age >= 65: add("over_65", {"age": age})
    if age is not None and age < 18: add("under_18", {"age": age})

    status = "ready_to_submit"
    if not ex.get("readable", True) or not texas or not nm or not signer_ok or low_conf or (age is not None and age < 18):
        status = "needs_review"
    elif not addr:
        status = "needs_dl_update"
    return Validation(status, addr, nm, texas, expired, age, age is not None and age >= 65, findings)
